### PR TITLE
✨(cli) add the new "auth" command

### PR DESCRIPTION
**Pull request by jmaupetit**
## Purpose

Generating credentials to authenticate to our LRS should be smooth and painless for our users.

## Proposal

- [x] add a new `auth` command to ease new user credentials generation
- [x] add pydantic models for user credentials checking and credentials file validation
- [x] track authentication issues to the LRS server

### CODE REVIEW

**Comment by jmaupetit** [comment]
I'll have to bypass the git lint job on this PR as we are removing a `print` statement from the documentation :sweat_smile:

**Review by quitterie-lcs — COMMENTED** [review]
:tada: :muscle:

File: src/ralph/cli.py [hunk 1]
```diff
@@ -177,6 +180,80 @@ def wrapper(command):
...
+        auth_file.touch()
+
+        try:
+            users = json.loads(auth_file.read_text(encoding="utf-8"))
+        except json.JSONDecodeError:
+            logger.warning(
+                "Authentication file <%s> is empty or contains invalid JSON data!",
+                settings.AUTH_FILE,
+            )
+            users = []
```
[the remark this answers is not in the document]
Comment (jmaupetit) [reply src/ralph/cli.py]: Well spotted. I've changed this: the credentials file is left unchanged if it's badly formatted.

File: setup.cfg [hunk 2]
```diff
@@ -53,6 +53,7 @@ cli =
     sentry-sdk[fastapi]>=1.9.0
 dev =
     bandit==1.7.4
+    bcrypt==4.0.1
```
Comment (SergioSim) [inline setup.cfg:56]: Isn't `bcrypt` a `cli` dependency as `src/ralph/cli.py` uses it?

Comment (jmaupetit) [reply setup.cfg:56]: Well spotted! It should be. I thought I did this. I bet it's a `git rebase` issue :roll_eyes:

**Review by quitterie-lcs — APPROVED** [review]

File: src/ralph/cli.py [hunk 3]
```diff
@@ -177,6 +180,83 @@ def wrapper(command):
...
+            try:
+                users = json.loads(auth_file_content)
+            except json.JSONDecodeError as exc:
+                logger.error(
+                    "Authentication file <%s> contains invalid JSON data!",
+                    settings.AUTH_FILE,
+                )
+                raise exc
+
+        users.append(credentials)
```
Comment (SergioSim) [inline src/ralph/cli.py:247]:
```suggestion
        users.insert(0, credentials)
```

By inserting the new user at the top of the list we let the user to redefine his password (as the LRS only takes the first matching username occurrence for authentication). WDYT?

File: src/ralph/cli.py [hunk 4]
```diff
@@ -177,6 +180,83 @@ def wrapper(command):
...
+
+    if write:
+        logger.info("Will append new credentials to: %s", settings.AUTH_FILE)
+
+        # Force Path object instantiation so that the file creation can be
+        # faked in a test environment.
+        auth_file = Path(settings.AUTH_FILE)
+        # Create the authentication file if it does not exist
+        auth_file.touch()
+        auth_file_content = auth_file.read_text(encoding="utf-8")
```
Comment (SergioSim) [inline src/ralph/cli.py:234]:
```suggestion
        auth_file_content = auth_file.read_text(encoding=settings.LOCALE_ENCODING)
```

Same suggestion for the write encoding

File: src/ralph/cli.py (hunk 3 above)
Comment (jmaupetit) [reply src/ralph/cli.py:247]: That's seems convoluted and not straightforward. What about validating the credentials file instead? (_i.e._ ensuring that the username field value is unique).

Comment (SergioSim) [reply src/ralph/cli.py:247]: I agree) With validation, it would be much better)

Comment (jmaupetit) [reply src/ralph/cli.py]: I've added a fixup commit that implements Pydantic models for credentials file validation. WDYT?

File: src/ralph/cli.py [hunk 5]
```diff
@@ -177,6 +181,80 @@ def wrapper(command):
...
+        # faked in a test environment.
+        auth_file = Path(settings.AUTH_FILE)
+        # Create the authentication file if it does not exist
+        auth_file.parent.mkdir(parents=True, exist_ok=True)
+        auth_file.touch()
+
+        users = ServerUsersCredentials.parse_obj([])
+        # Parse credentials file if not empty
+        if auth_file.stat().st_size:
+            users = ServerUsersCredentials.parse_file(auth_file)
```
[the remark this answers is not in the document]
Comment (jmaupetit) [reply src/ralph/cli.py:240]: I asked myself the same question and concluded that Pydantic is that much explicit in its validation exceptions that adding some more noise around is not required. WDYT?

File: src/ralph/cli.py [hunk 6]
```diff
@@ -23,6 +26,7 @@
 from click_option_group import optgroup
 from pydantic import BaseModel
 
+from ralph.api.auth import ServerUsersCredentials, UserCredentials
```
[the remark this answers is not in the document]
Comment (jmaupetit) [reply src/ralph/cli.py]: I used the `import_string` utility.

**Comment by jmaupetit** [comment]
Aaaaand almost done. Keep it up! Ping @SergioSim

**Review by quitterie-lcs — COMMENTED** [review]
A few remarks

File: setup.cfg [hunk 7]
```diff
@@ -48,6 +48,7 @@ backend-swift =
 backend-ws =
     websockets>=10.3
 cli =
+    bcrypt==4.0.1
```
[the remark this answers is not in the document]
Comment (jmaupetit) [reply setup.cfg]: Good point. I've introduced some more flexibility compatible with `lrs` optional dependencies.

**Review by SergioSim — APPROVED** [review]
LGTM) :+1: 

image

**Pull request merged**